plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/repository/AbstractSourceAttachmentDownloaderJob.java

```diff
@@ -0,0 +1,142 @@
+package com.google.cloud.tools.eclipse.appengine.libraries.repository;
+
+import com.google.cloud.tools.eclipse.appengine.libraries.LibraryClasspathContainer;
+import com.google.cloud.tools.eclipse.appengine.libraries.persistence.LibraryClasspathContainerSerializer;
+import com.google.cloud.tools.eclipse.util.status.StatusUtil;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.jdt.core.IClasspathContainer;
+import org.eclipse.jdt.core.IClasspathEntry;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.core.JavaModelException;
+
+public abstract class AbstractSourceAttachmentDownloaderJob extends Job {
+
+  private long retryDelayFactor = 3000L;
+  private int retryAttempts = 10;
+
+  @Inject
+  private IJavaProject javaProject;
+  @Inject
+  @Named("classpathEntryPath")
+  private IPath classpathEntryPath;
+  @Inject
+  private LibraryClasspathContainerSerializer serializer;
+
+  public AbstractSourceAttachmentDownloaderJob(String name) {
```

Comment:
The `getSourcePath` method (probably inaptly named) supposed to produce the source path which can happen either via downloading the artifact or simply returning the path if the artifact is already downloaded. This decision is transparent to the caller in case of M2E (i.e. M2E will download the artifact if not found locally).

The retry is triggered from the `setSourceAttachmentPath` method for the reasons described in the above comment thread: I noticed that not all containers are initialized by the time it's called and running it again a bit later gives the opportunity to the initializer to create the container. Again, I'll revisit this process to avoid relying on the retry mechanism to fix this problem.